### ISSUE TITLE
Relative/absolute filenames in buffers

### ISSUE THREAD

**Issue by memeplex** [neovim/neovim#3764]
vim names its file buffers according to the path passed to `:e`. When the passed path is relative the buffer name becomes relative, when it's absolute the buffer name becomes absolute. But then `:cd X` and `:lcd X` rename the buffers to become absolute if outside X, otherwise relative. This behavior is not consistent and thus confusing: sometimes absolute means outside the current working directory, sometimes not. For quickfix, which parses the output of external command, this sometimes implies that files in the current directory are shown with absolute paths (in case the external command happens to report the full path).

This was discussed a time ago in vim mailing list but never really addressed:   

https://groups.google.com/forum/#!topic/vim_use/Vq0z2DJH2So

(And see my awful workaround at the end: `0split | lcd . | quit`, I would like to remove it from my config for good!)

There are two issues described there:

[] Consistently set buffer names. The obvious rule is: relative if inside current working directory, otherwise absolute.
[] I would be nice if `getcwd()` could specify if the desired cwd is the local or the global one.

**Comment by memeplex** [comment]
If you agree this is a bug I will hack a patch and open a PR. Otherwise I will manage to live with my current workaround.

**Comment by ZyX-I** [comment]
Where does this make a difference? I can only imagine having a difference in a status line, which is fixed by proper `&statusline` setting.

**Comment by memeplex** [comment]
quickfix tend to open a number of files inside the "current project" and the buffer list becomes quickly cluttered.

Automation becomes harder also, as it's not possible to depend on a regular, simple behavior (in fact, that's how I hit this).

To recap: inconsistency and clutter.

**Comment by ZyX-I** [comment]
Did not saw anything like this. Though I have an autocommand in my vimrc which upon switching to a window changes its directory to the project root (determined by presence of the VCS-specific directories like .hg) or file directory (in case nothing like this was found), effectively doing the same thing your workaround does I think.

**Comment by memeplex** [comment]
That approach is not without problems. Auto-calling `cd` on window change would modify a global setting, while auto-calling `lcd` would establish a local setting. In general, both alternatives are wrong, as they impose a scope (global or local) for the setting which may be undesirable for the case being. Moreover, it's not possible to honour the current window cwd scope (global or local) as it's not possible to know whether the current working directory was globally or locally set by the user.

That's why my workaround open a discardable window to execute `lcd`, and that's why it's so ugly.

**Comment by ZyX-I** [comment]
@memeplex Did you try using `:file` in a workaround?

**Comment by justinmk** [comment]
This caused me lot of pain when building dirvish.vim. `:file` is not a great solution. It also makes it very difficult to determine whether a resolved buffer exists. 

It makes ZERO sense for a buffer name to be anything except absolute path. Pretty/relative name is a matter of configuring the status line.

**Comment by memeplex** [comment]
The "always absolute" criterion is the most coherent one, except that vim uses the buffer name both as a UI thing and as a filename. The "always absolute" criterion will thus imply that everywhere the buffer name could be shown, a decision must be taken about how to show it. One place is the modeline. Another place is the buffer list (I wouldn't like to get long absoulte paths for files in the current working directory there). There will be other places, for sure. "Always absolute" is logically more sound, as it separates different concerns, but could trigger the need for a number of new format options (how to show the buffer name here and there). "Relative inside the cwd/Absolute outside the cwd" could be a simpler to implement compromise (I guess it's just a matter of normalizing the name on file buffer creation), as it's what vim/nvim already poorly does.

**Comment by azure-satellite** [comment]
+1 on always absolute. I see 2 pros:

- We don't know how people will like to format their paths in the UI. It makes more sense to make all context available.
- Easier logic in neovim core (ex: comparing if two paths are the same)

Also. Sometimes the cwd doesn't matter. Very often I open files from here and there and I don't care if one of those is under my cwd.

**Comment by ZyX-I** [comment]
I would vote for “always relative if under the current dir” if not the fact that places where this matters (powerline statusline and command-t for switching buffers) are ignoring what Vim thinks is a buffer name and use their own means of deciding what exactly to show.

AFAIR I would not welcome this change though when I was using `:b` for switching buffers in place of command-t.

> Easier logic in neovim core (ex: comparing if two paths are the same)

Where have you seen this? Buffers have `b_ffname` attached which contain absolute path and if code deals with `buf_T` structures it may simply use `b_ffname`.

**Comment by azure-satellite** [comment]
> Where have you seen this? Buffers have b_ffname attached which contain absolute path and if code deals with buf_T structures it may simply use b_ffname.

@ZyX-I Sorry you definitely have much more domain knowledge on vim's codebase. I was just talking in the abstract

**Comment by ZyX-I** [comment]
@azure-satellite I cannot say that I know all places where file names are compared, so you had some chances to prove yourself. But what I do know is the existence of `b_ffname` `buf_T` structure attribute which makes these chances unlikely and one of the points in defence of your position invalid.

By the way, one thing absolute paths: Vim tends to use `~/`-style paths in many places and sometimes even compare *these* paths: e.g. replacing ~absolute path with /absolute path was advertised as one of the advantages of ShaDa format (which also simplified some code). I.e. buffer had `b_ffname` attribute, but viminfo file for some reason contained ~absolute path and thus determining one of the global marks belongs to one of the existing buffers involved iterating over buffers, using `home_replace_save` on each `b_ffname` (which turns /absolute to ~absolute) and only then comparing the paths for equality (do not ask me why instead of doing this ~absolute path was not expanded to /absolute). I simply use plain `b_ffname` in ShaDa code because this is clearer then messing with ~absolute paths in any fashion, even in saner one. ~absolute paths are still used in a number of places: e.g. in swap file headers.

Thus even if absolute paths are the way to go, there is still a question whether they need to be ~absolute when possible.

---

~absolute is short-hand for “`~/`-style absolute path” (example: `~/.config/nvim`).

/absolute is a short-hand for “fully expanded absolute path” (example: `/home/zyx/.config/nvim`).

**Comment by justinmk** [comment]
Only concrete issue I see here is this:

> or quickfix, which parses the output of external command, this often implies that files in the current directory are shown with absolute paths

 
See also https://github.com/vim/vim/issues/549.

**Comment by memeplex** [comment]
As I see it, quickfix is consistent with the arguably inconsistent file buffer naming rule: create a buffer with the name given as is, then relativize it upon any call to cd or lcd in any window. I would expect that either the name is absolute forever when the buffer is created that way, either the name is absolute only when outside the current working directory. The current behavior is a confusing mix of those alternatives. "Fixing" quickfix would be ok but ad hoc.

**Comment by justinmk** [comment]
> I would expect that either the name is absolute forever when the buffer is created that way

But that wouldn't fix the quickfix issue, unless I'm missing something.

**Comment by memeplex** [comment]
No, it wouldn't. That's way my proposal was to follow the relative inside cwd, absolute elsewhere rule. vim/neovim already follows this rule except at buffer creation. That's what I find inconsistent. But other people here prefer the always absoulte rule. Nevertheless, I think this last rule may have to introduce a distinction between the buffer name and its presentation, which is currently lacking.

**Comment by hiberabyss** [comment]
Suffering from this problem.

**Comment by lacygoill** [comment]
I'm not sure this is relevant to this issue, but a patch ([8.0.1787](https://github.com/vim/vim/commit/a796d46f29e3cc235cc981696d7ee80faccb5000)) was recently merged in Vim, whose purpose seems to be to always shorten the filepaths (making them relative to the current working directory when they're below the latter, I suppose).

**Comment by P4Cu** [comment]
I've also felt that's annoying:
https://github.com/neovim/neovim/pull/8967
